### DISCUSSION TITLE
Offline playback from mobile app

### DISCUSSION THREAD

**Discussion by befuddled-dude** [advplyr/audiobookshelf#874] [General]
I just started playing with Audiobookshelf, and it seems like a good start for what I'm looking for.

I have an Android phone that has a metered mobile data plan.  Ideally, I'd like to be able to:

1. Download audio books from the server to local storage 
2. When not on a wifi network, I'd like the app to use the local copy of the audio files for playback
3. Sync the progress back to the server with any available network
4. Have an app setting to disallow mobile data generally for playback 

This would allow me to switch between devices for playback (wifi iPad, desktop, phone) but not rack up streaming costs with my phone.

Unfortunately, the Android app is a bit unstable for me. I was able use the app to download a 'book' from the server to my phone, but when I try to play that book, the app crashes.  I also can't seem delete the downloaded book folder via the app (Local Media->Local Folders->{FolderName}->{BookName}->Remove & Delete Files.) I get a red "Failed to Delete" message pop up.

Obviously Item 1 above is in place, but are the others ones in place or planned?

I'll be happy to create an issue for the bugs that I see here, but I wasn't sure if that's the right place for the Android app.  Let me know where, and I do what I can to provide details, logs, etc.

Thanks!

**Comment by advplyr** [comment]
What version of the app?

**Reply by befuddled-dude** [reply]
Android app is 0.9.55-beta, and the server is  v2.1.1

New test:
I removed the downloaded book from the "Local Library Items", and manually deleted the actual folder and files from the phone using a file manager.  Then I tried to download the book again (it's 6 mp3 files) using the app.  This time it seemed to download the files, and after it got to 100% it gave a red message pop-up "Item download complete but failed to create library item". I couldn't seem to download any other books either.

Then I used the app to delete the local base folder for books, and tried the book download again.  It succeeded, and I was able to do local playback when offline.  I guess something got broke between betas?

Next, I turned on cell connectivity, but not wifi, and I see now that for my downloaded book the distinction is made by having a green "Play" button, instead of the green "Stream" button for books only on the server.  ( **Suggestion**: maybe make it more explicit to something like "Local Playback", or maybe a small bit of text under the play button.)

Once I'm playing the local book, I do see on the player screen the green "LOCAL" across the top, which is great.

Other observations/suggestions/requests:
- **Suggestion**: Optionally allow syncing of local playback progress over metered and non-metered connections.
- **Weird Possible Bug**: I don't know how it happened, but at one point when experimenting with the playback of the locally downloaded book when on the mobile network, it somehow started streaming instead, and I saw a "DIRECT" on top of the player screen.  I can't seen to recreate that condition.  
- **Suggestion**: Based on the above item, I'd suggest an option to globally disable streaming when on a metered network.

Thanks for creating this awesome app and server!

**Reply by advplyr** [reply]
If you play the audiobook or have it playing already before it has finished the download & scan then it will stream.
I think this was brought up in discord before so there isn't a github issue for it which is probably why I keep forgetting to do it but this will need to be updated to switch the playback to local once a download completes.

**Reply by Binocularbath** [reply]
Has there been any progress on this

**Reply by ansible42** [reply]
@Binocularbath it works on my setup with the android app. But the initial download it manual.
